mlflow/dspy/callback.py

```diff
@@ -0,0 +1,190 @@
+        }
+
+        self._start_span(
+            call_id,
+            name=f"{instance.__class__.__name__}.__call__",
+            span_type=span_type,
+            inputs=self._unpack_kwargs(inputs),
+            attributes=attributes,
+        )
+
+    def on_lm_end(
+        self, call_id: str, outputs: Optional[Any], exception: Optional[Exception] = None
+    ):
+        self._end_span(call_id, outputs, exception)
+
+    def on_adapter_format_start(self, call_id: str, instance: Any, inputs: Dict[str, Any]):
+        self._start_span(
+            call_id,
+            name=f"{instance.__class__.__name__}.format",
+            span_type=SpanType.PARSER,
+            inputs=self._unpack_kwargs(inputs),
+            attributes={},
+        )
+
+    def on_adapter_format_end(
+        self, call_id: str, outputs: Optional[Any], exception: Optional[Exception] = None
+    ):
+        self._end_span(call_id, outputs, exception)
+
+    def on_adapter_parse_start(self, call_id: str, instance: Any, inputs: Dict[str, Any]):
+        self._start_span(
+            call_id,
+            name=f"{instance.__class__.__name__}.parse",
+            span_type=SpanType.PARSER,
+            inputs=self._unpack_kwargs(inputs),
+            attributes={},
+        )
+
+    def on_adapter_parse_end(
+        self, call_id: str, outputs: Optional[Any], exception: Optional[Exception] = None
+    ):
+        self._end_span(call_id, outputs, exception)
+
+    def _start_span(
+        self,
+        call_id: str,
+        name: str,
+        span_type: SpanType,
+        inputs: Dict[str, Any],
+        attributes: Dict[str, Any],
+    ):
+        # Get parent span in this order:
+        # 1. If there is an parent component in DSPy, use its span as parent span.
+        # 2. If there is an active span in MLflow, use it as parent span.
+        # 3. Otherwise, start a new root span.
+        if parent_call_id := ACTIVE_CALL_ID.get():
+            parent_span = self._call_id_to_span.get(parent_call_id)
+            if not parent_span:
+                _logger.warning("Failed to create a span. Parent span not found.")
+        elif active_span := mlflow.get_current_active_span():
+            parent_span = active_span
+        else:
+            parent_span = None
+
+        common_params = {
+            "name": name,
+            "span_type": span_type,
+            "inputs": inputs,
+            "attributes": attributes,
+        }
+
+        if parent_span:
+            span = self._client.start_span(
+                request_id=parent_span.request_id, parent_id=parent_span.span_id, **common_params
+            )
+        else:
+            span = self._client.start_trace(**common_params)
+
+        self._call_id_to_span[call_id] = span
+
+    def _end_span(
+        self,
+        call_id: str,
+        outputs: Optional[Any],
+        exception: Optional[Exception] = None,
+    ):
+        span = self._call_id_to_span.pop(call_id)
```

Comment:
To confirm, `call_id` is always present?

Reply:
Theoretically yes, but cannot guarantee. We can add a guardrail here.